resolver.go

```diff
@@ -145,6 +145,14 @@ func (r *resolver) Start() error {
 		return fmt.Errorf("setting up IP table rules failed: %v", err)
 	}
 
+	// If the container goes down quickly after start resolver.Stop() can
+	// get called before Start() completes. Check if the socket is still
+	// valid before Activating the DNS server. This avoids a panic in dns
+	// libarary code because of missing nil check for interface's data.
+	// github.com/docker/docker/issues/28112
+	if r.conn == nil {
```

Comment:
This reduces the chances for the bug but does not eliminate them. We need to Lock the resolver before accessing/modifying its members. And we need to pass a reference of `r.conn` once acquired under lock.

Also, wouldn't the same issue be there also when we construct the tcpServer ?